identity-did/src/verifiable/document_verifier.rs

```diff
@@ -36,54 +38,87 @@ use crate::Result;
 // Document Verifier - Simplifying Digital Signature Verification Since 2021
 // =============================================================================
 
-pub struct DocumentVerifier<'base, T = Object, U = Object, V = Object> {
+pub struct DocumentVerifier<'base, 'options, T = Object, U = Object, V = Object> {
   document: &'base CoreDocument<T, U, V>,
-  options: VerifierOptions,
+  options: Cow<'options, VerifierOptions>,
 }
 
-impl<'base, T, U, V> DocumentVerifier<'base, T, U, V> {
+impl<'base, T, U, V> DocumentVerifier<'base, 'static, T, U, V> {
   pub fn new(document: &'base CoreDocument<T, U, V>) -> Self {
     Self {
       document,
-      options: VerifierOptions::default(),
+      options: Cow::Owned(VerifierOptions::default()),
     }
   }
+}
 
+impl<'base, 'options, T, U, V> DocumentVerifier<'base, 'options, T, U, V> {
   /// Overwrites the [`VerifierOptions`].
   #[must_use]
-  pub fn options(mut self, options: VerifierOptions) -> Self {
+  pub fn options<'new_options>(
+    self,
+    options: &'new_options VerifierOptions,
+  ) -> DocumentVerifier<'base, 'new_options, T, U, V> {
+    /*
     self.options = options;
     self
+    */
+    DocumentVerifier {
+      document: self.document,
+      options: Cow::Borrowed(options),
+    }
   }
 
   /// Verify the signing verification method relationship matches this.
   ///
   /// NOTE: `purpose` overrides the `method_scope` option.
   #[must_use]
-  pub fn method_scope(mut self, method_scope: MethodScope) -> Self {
-    self.options = self.options.method_scope(method_scope);
+  pub fn method_scope(self, method_scope: MethodScope) -> DocumentVerifier<'base, 'static, T, U, V> {
+    /*self.options = self.options.method_scope(method_scope);
     self
+    */
+    DocumentVerifier {
+      document: self.document,
+      options: Cow::Owned(self.options.into_owned().method_scope(method_scope)),
+    }
   }
```

Comment:
Maybe instead of all these setters on `DocumentVerifier` we can just leave it as an immutable borrow field and allow overwriting that borrow as a whole with `DocumentVerifier::set_options`? I don't imagine many developers will use `DocumentVerifier` over `IotaDocument::verify_data`.

The conversion of `Cow` to owned is unexpected in my opinion.

Reply:
Thanks for your early feedback :+1:  

> The conversion of Cow to owned is unexpected in my opinion.

The fact that it must convert the `Cow`to owned is somewhat reflected in the function signature (the return type has a `'static` lifetime in the second generic parameter). To make this even more obvious we could introduce a sealed trait 
`trait OptionsState` that is implemented for two zero sized types `Borrowed` and `Owned`. Then `DocumentVerifier` is changed to  
```
DocumentVerifier<'base, 'options, State: VerifierOptionState, T = Object, U = Object, V = Object,>
```
and then there is an implementation 
```
impl DocumentVerifier<'base, 'options, Borrowed, T, U, V> {
     fn into_owned(self) -> DocumentVerifier<'base, 'static, Owned, T, U, V>; 
}
```
and the methods that change the `options` are only implemented on `DocumentVerifier<'base, 'static, Owned, T, U, V>`. This would make the fact that one needs to clone if one wants to manipulate the verifier options directly on `DocumentVerifier` very obvious. The question is whether it is worth the extra code? Otherwise it might be OK to not let `DocumentVerifier` mutate the individual fields of its verifier options as you mentioned.

Reply:
Decided to refactor `DocumentVerifier` to take the verifier options in the `verify` method instead of having them as a field in the struct.